src/OpenTelemetry.Instrumentation.Dependencies/Implementation/GrpcClientDiagnosticListener.cs

```diff
@@ -0,0 +1,94 @@
+﻿// <copyright file="GrpcClientDiagnosticListener.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Net.Http;
+using System.Text.RegularExpressions;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Instrumentation.Dependencies.Implementation
+{
+    internal class GrpcClientDiagnosticListener : ListenerHandler
+    {
+        private static readonly Regex GrpcMethodRegex = new Regex(@"((?<package>\w+)\.)?(?<service>\w+)/(?<method>\w+)", RegexOptions.Compiled);
+
+        private readonly ActivitySourceAdapter activitySource;
+        private readonly PropertyFetcher startRequestFetcher = new PropertyFetcher("Request");
+
+        public GrpcClientDiagnosticListener(ActivitySourceAdapter activitySource)
+            : base("Grpc.Net.Client", null)
+        {
+            this.activitySource = activitySource;
+        }
+
+        public override void OnStartActivity(Activity activity, object payload)
+        {
+            const string EventNameSuffix = ".OnStartActivity";
+            if (!(this.startRequestFetcher.Fetch(payload) is HttpRequestMessage request))
+            {
+                InstrumentationEventSource.Log.NullPayload(nameof(GrpcClientDiagnosticListener) + EventNameSuffix);
+                return;
+            }
+
+            var grpcMethodTag = activity.Tags.FirstOrDefault(tag => tag.Key == "grpc.method");
+            var grpcMethod = grpcMethodTag.Value?.Trim('/');
+
+            // TODO: Avoid the reflection hack once .NET ships new Activity with Kind settable.
+            activity.GetType().GetProperty("Kind").SetValue(activity, ActivityKind.Client);
+            activity.DisplayName = grpcMethod;
+
+            this.activitySource.Start(activity);
+
+            if (activity.IsAllDataRequested)
+            {
+                var rpcService = GrpcMethodRegex.Match(grpcMethod).Groups["service"].Value;
```

Comment:
What happens if regex doesn't match `grpcMethod`? Right now it will just error.

Options:
1. Don't add `rpc.service` tag
2. Or add the tag with "unknown" value.
3. Or add the tag with grpcMethod value.

I like option 3.

Reply:
Currently, if the regex doesn't match then this resolves to an empty string... not ideal.

The `rpc.service` tag is listed in the spec as not required but recommended: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/rpc.md#attributes.

Options 1 or 2 seem ok, but since option 3 also includes the method name it seems to go against the definition of the attribute.

Reply:
Ok, 1 is a good choice. To be honest writing an empty string is fine too. As long as it doesn't error on an unexpected value.

Looks like the spec now says `rpc.service` is package + service. And `rpc.method` is newly added.

Reply:
Ah, good catch I missed the newly added `rpc.method` attribute. This has been added. I've also changed it so that if the regex match fails then `rpc.service` and `rpc.method` attributes do not get added.